### PR TITLE
Refactor to enable support for Amp in PyTorch and better support future phase2 work

### PR DESCRIPTION
**Pull request by markurtz**

### CODE REVIEW

File: src/sparseml/pytorch/optim/manager.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/manager.py:292]: nice

File: src/sparseml/optim/modifier.py
Comment (bfineran) [inline src/sparseml/optim/modifier.py:347]: should get a test for this in. am ok with later as tech debt since this req doesn't seem immediately breaking

File: src/sparseml/pytorch/optim/modifier_pruning.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:162]: should check `strict` here

File: src/sparseml/pytorch/optim/modifier.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier.py:147]: instead of `pass` I think the minimal functionality here should be checking for `strict` and raising an index error if anything is present.

File: src/sparseml/pytorch/optim/manager.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/manager.py:37]: nit: not sure if this name makes it clear what the class' purpose is. could we change the name to something like `TrainingStepRecipeWrapper` to singal that we're wrapping things that aren't just solo optimizers?

File: src/sparseml/pytorch/optim/manager.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/manager.py:368]: can we call into initialize instead of failing here? an implicit call to initialize would save an extra line in standard integrations.

File: src/sparseml/pytorch/optim/modifier_pruning.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:82]: would it make sense to add a `finalize` method that checks if the pruning masks are enabled, and if so, calls `apply` one last time and then disables them?

Comment (bfineran) [reply src/sparseml/pytorch/optim/modifier_pruning.py:82]: looks like you read my mind

File: src/sparseml/pytorch/optim/modifier_pruning.py
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:259]: let's call `apply()` one more time if the masks are enabled

**Review by bfineran — CHANGES_REQUESTED** [review]
Looks overall good to me. Thanks for getting this bug resolved while also improving the codebase. Since `ScheduledOptimizer` is deprecated, could we update one of our examples to use these changes or update this PR with a brief example of the new flow for AMP?

Also are there any docs that need updating?

File: src/sparseml/optim/modifier.py
Comment (markurtz) [reply src/sparseml/optim/modifier.py:347]: Agree on the test, but let's push that off for later -- there are some integration tests we'll need to do between pruning and quantization to make sure they order correctly based on this

File: src/sparseml/pytorch/optim/manager.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/manager.py:37]: Will rename to RecipeManagerStepWrapper in prep for renaming the Managers to RecipeManager for phase2

File: src/sparseml/pytorch/optim/manager.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/manager.py:368]: yes, will default to calling into initialize if not called. Will also work out well for TF and keras

File: src/sparseml/pytorch/optim/modifier.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/modifier.py:147]: good point, fixing

File: src/sparseml/pytorch/optim/modifier_pruning.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/modifier_pruning.py:82]: :)

File: src/sparseml/pytorch/optim/modifier_pruning.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/modifier_pruning.py:162]: Because of the way we load the masks, strict is the only option, otherwise, we'll have to fail. I put a note in the docs for this

File: src/sparseml/pytorch/optim/modifier_pruning.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/modifier_pruning.py:259]: done

**Comment by markurtz** [comment]
> Looks overall good to me. Thanks for getting this bug resolved while also improving the codebase. Since `ScheduledOptimizer` is deprecated, could we update one of our examples to use these changes or update this PR with a brief example of the new flow for AMP?
> 
> Also are there any docs that need updating?

I'm holding off on changing over examples/docs for a minor point release. So, trying to keep everything originally functioning and enabling this silently on the backend for the ultralytics integrations to unblock those for now. On 0.4 we'll update examples and docs

**Review by bfineran — DISMISSED** [review]

File: src/sparseml/pytorch/optim/manager.py
Comment (natuan) [inline src/sparseml/pytorch/optim/manager.py:175]: Can you clarify what the difference between the self._wrapped and the self._wrapped_optimizer? In the context of amp support, what would we pass in for the _wrapped?

File: src/sparseml/optim/modifier.py
Comment (natuan) [inline src/sparseml/optim/modifier.py:423]: based on identifiers

File: src/sparseml/pytorch/optim/manager.py
Comment (markurtz) [reply src/sparseml/pytorch/optim/manager.py:175]: Wrapped is what we override the step call for. Wrapped optimizer is the pytorch optimizer implementation needed to be passed through to the modifiers

Comment (natuan) [reply src/sparseml/pytorch/optim/manager.py:175]: I saw we passed the same optimizer into both the _wrapped and _wrapped_optimizer (in optimizer.py). This is not always the case, right? (otherwise, we could just use _wrapped_optimizer)

**Review by natuan — APPROVED** [review]

**Pull request merged**
